network-config-analyzer/MinimizeFWRules.py

```diff
@@ -0,0 +1,754 @@
+            self._create_initial_fw_rules_from_base_elements_list(self.peer_pairs_with_partial_ns_expr))
+        return initial_fw_rules
+
+    def _add_all_fw_rules(self):
+        """
+        Computation of fw-rules, following the ns-grouping of peer_pairs.
+        Results are at: self.minimized_rules_set
+        :return: None
+        """
+        # create initial fw-rules from ns_pairs, peer_pairs_with_partial_ns_expr, peer_pairs_without_ns_expr
+        initial_fw_rules = self._create_all_initial_fw_rules()
+        # option1 - start computation when src is fixed at first iteration, and merge applies to dst
+        option1, convergence_iteration_1 = self._create_merged_rules_set(True, initial_fw_rules)
+        # option2 - start computation when dst is fixed at first iteration, and merge applies to src
+        option2, convergence_iteration_2 = self._create_merged_rules_set(False, initial_fw_rules)
+
+        # self.post_processing_fw_rules(option1)
+        # self.post_processing_fw_rules(option2)
+
+        if self.output_config.fwRulesRunInTestMode:
+            equiv1 = self.check_peer_pairs_equivalence(option1)
+            equiv2 = self.check_peer_pairs_equivalence(option2)
+            assert equiv1
+            assert equiv2
+            # add info for documentation about computation results
+            self.results_info_per_option['option1_len'] = len(option1)
+            self.results_info_per_option['option2_len'] = len(option2)
+            self.results_info_per_option['convergence_iteration_1'] = convergence_iteration_1
+            self.results_info_per_option['convergence_iteration_2'] = convergence_iteration_2
+            self.results_info_per_option['equiv1'] = equiv1
+            self.results_info_per_option['equiv2'] = equiv2
+
+        if self.output_config.fwRulesDebug:
+            print('option 1 rules:')
+            self._print_firewall_rules(option1)
+            print('option 2 rules: ')
+            self._print_firewall_rules(option2)
+
+        # choose the option with less fw-rules
+        if len(option1) < len(option2):
+            self.minimized_fw_rules.extend(option1)
+            return
+        self.minimized_fw_rules.extend(option2)
+
+    def _get_grouping_result(self, fixed_elem, set_for_grouping_elems, src_first):
+        """
+        Apply grouping for a set of elements to create grouped fw-rules
+        :param fixed_elem: the fixed elements from the original fw-rules
+        :param set_for_grouping_elems: the set of elements to be grouped
+        :param src_first: a bool flag to indicate if fixed_elem is src or dst
+        :return: A list of fw-rules after possible grouping operations
+        """
+        res = []
+        fw_rule_elem_type = type(FWRuleElement([]))
+        # partition set_for_grouping_elems into: (1) ns_elems, (2) pod_and_pod_labels_elems, (3) ip_block_elems
+        ns_elems = set(elem for elem in set_for_grouping_elems if type(elem) == fw_rule_elem_type)
```

Comment:
I don't really understand the trick with `type()` here. Maybe a comment why it is required?

Reply:
changed so that it is not required.